### ISSUE TITLE
3rd-party packages cannot be resolved when specifying files on command line

### ISSUE THREAD

**Issue by BobReid** [microsoft/pyright#1015]
**Describe the bug**
I am in the process of integrating pyright into our code review process by implementing a custom [pronto](https://github.com/prontolabs/pronto) runner. Pronto works by detecting the github branch, pulling the diff between the branch and master, and passing the patch data to the extensions to do the actual analysis. 

Ideally you only need to analyze the files that have changed in the PR since it is much quicker than running on the entire project. In my case ~1 second vs ~35 seconds for pyright. This approach is does not work with pyright because it incorrectly reports 3rd party packages as missing imports. 

I am not sure extactly why this is happening but my intuition is that it is because you cannot properly specify the venv folder from the command line. 

The -v (--venv-path) command line option is equivalent to the ```venvPath``` configuration file parameter but there is no equivalent to the ```venv``` configuration file parameter. Since the confirgation file is ignored I cannot specify the ```venv```. I have tried setting the -v option to invclude the venv folder itself (ie. './.venv' ) but that does not work either. 

**To Reproduce**
1. Create project with a venv folder in the project directory
2. Run ```pyright --lib {FILE}``` where FILE is a python file that contains a 3rd party import and has not errors reported by pylance (or when running pyright against the entire project with venv

**Expected behavior**
1. FILE is reported as having no errors on the command line

**VS Code extension or command-line**
Command Line
Version: pyright 1.1.69 

**Additional context**
| Run Mode | venv set in config | Errors Reported |
------------|-------------------|----------|
Project | Yes | No |
Project | No | Yes |
File | n/a | Yes

**Comment by erictraut** [comment]
The `venvPath` should point to a directory that contains one or more subdirectires, each with its own venv. The name of the venv to use for the workspace should be specified in the pyrightconfig.json file within the "venv" setting. In effect, Pyright will look in the directory `<venvPath>/<venv>`. The reason these are separated this way is to allow each developer on a project to define their own local `venvPath` while allowing everyone to share the same `venv` name for the project. 

The configuration file also allows you to specify different `venv` names for different subdirectories within your workspace. This is especially useful for large monorepos that contain many Python scripts and applications, each with their own configuration requirements.

If you would prefer not to use `venvPath` or `venv` configurations, you can omit these and configure the python interpreter within your shell environment. If no `venvPath" is specified, Pyright will retrieve the import resolution paths from the configured python interpreter (by executing "python3" or "python" and running a small script that outputs the paths in `sys.path`).

Refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/configuration.md) for more details.

You said "Since the confirgation file is ignored...". Which configuration file are you talking about? The "pyrightconfig.json" file will be honored by both the CLI version of pyright and the VS Code version of Pyright and Pylance.

**Comment by BobReid** [comment]
Thanks for the additional information. 

Our project is using Pipenv for the package management and we are specifically using the PIPENV_VENV_IN_PROJECT environment variable so the .venv folder lives in the package directory, not the python directory. 

I was able to work around it by wrapping pyright in ```pipenv run```, which causes the python environment to be loaded correctly when the python resolver script runs. It is non-obvious to have to wrap the node js executable in ```pipenv run```. 

I suppose this is not a bug at the moment since it is working as designed but I wonder in my opinion it would make sense to allow venv to be set on the command line along with venvPath to have more control in a CI pipeline.

**Comment by BobReid** [comment]
I will close this issue since it is working as designed but I am curious on the reasoning for not supporting the config file when using file paths on the command line. There are a ton of options that are not available on the command line when analyzing specific files. 

It would be useful to respect the config file and just override with anything specified on the command line.

**Comment by erictraut** [comment]
Pyright's configuration system is already very complicated because it needs to merge command-line options with config options with VS Code settings (at both the user and workspace levels) _and_ apply the appropriate defaults in all cases _and_ handle interactions between settings. It's gotten to the point where it's very difficult to reason about, and the code is hard to maintain, so I'm reluctant to add even more redundancy between CLI options and the config file.

The pyrightconfig.json config file is intended to checked in to the repo, so all configuration options within it should be common to all developers and the CI system. The command-line switches are mainly for settings that are unique to each machine. That's why "venvpath" is a CLI option but "venv" is not. Those are the principles I've been using to decide whether to expose settings on the command line. 

If there are specific use cases that cannot be handled with the current configuration mechanisms, let's discuss the specifics of those cases to see if there's a way we can accommodated without violating these principles.

**Comment by BobReid** [comment]
My argument was not for exposing more options to the command line. My suggestion is to allow the use of the config file when specifying files on the command line. 

In my own testing it seemed that nothing in the configuration file was picked up as long as I listed paths to analyze on the command line. I assume this is the intended behaviour because of the documentation on the -p flag stating that it doesn't work when specified files. 

I agree expanded on the command line options is not the ideal solution and its more complicated than necessary but I have basically no control over the output when analyzing speicific files. 

Ideally I have my configuration in the config file, then during CI I can pass in the files involved in the PR and have them analyzed with all settings in the config respected. This is the common approach for most static analysis tools that I use but it doesn't seem to be working. As soon as I set a file on the command line i ONLY get what is available on the command line. 

For example

I have a project with a single file test.py
```py
def foo():
    print('foo')
foo(1)
```

I am using the config file 
```json
{
  "reportGeneralTypeIssues": "warning"
}
```

When running the command ```pyright``` the type error is output as a warning as specified in the config.
```
Searching for source files
Found 1 source file
pyright_test/test.py
  5:5 - warning: Expected 0 positional arguments (reportGeneralTypeIssues)
0 errors, 1 warning 
Completed in 0.994sec
```

When running the command ```pyright test.py``` outputs the type error is output as an error (the default).
```
Searching for source files
Found 1 source file
pyright_test/test.py
  5:5 - error: Expected 0 positional arguments (reportGeneralTypeIssues)
1 error, 0 warnings 
Completed in 0.676sec
```

Therefore you have no control over the types of errors that are output when specifying files on the command line which is very limiting.

**Comment by erictraut** [comment]
Ah, thanks for the additional explanation. Makes sense.

**Comment by erictraut** [comment]
I've updated the CLI version of pyright to support config files along with file specs. The tricky part (and the reason I didn't originally allow this) is determining how the passed file specs should interact with the "include" and "exclude" sections of the config file. I decided that they would simply override the "include" section. All other aspects of the config file are honored. I've also updated the documentation to reflect this new behavior. This change will be in the next published version of pyright.

**Comment by BobReid** [comment]
Awesome, thanks!

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.71, which I just published.
